compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+          target.copy(dir = copyToDir(outDir, dir))
+      },
+      times, shouldDelete, threadLimit, shouldFail
+    )
+
+    /** Builds a `CompilationTest` which performs the compilation `i` times on
+     *  each target
+     */
+    def times(i: Int): CompilationTest =
+      new CompilationTest(targets, i, shouldDelete, threadLimit, shouldFail)
+
+    /** Builds a `Compilationtest` which passes the verbose flag and logs the
+     *  classpath
+     */
+    def verbose: CompilationTest = new CompilationTest(
+      targets.map(t => t.withFlags("-verbose", "-Ylog-classpath")),
+      times, shouldDelete, threadLimit, shouldFail
+    )
+
+    /** Builds a `CompilationTest` which keeps the generated output files
+     *
+     *  This is needed for tests like `tastyBootstrap` which relies on first
+     *  compiling a certain part of the project and then compiling a second
+     *  part which depends on the first
+     */
+    def keepOutput: CompilationTest =
+      new CompilationTest(targets, times, false, threadLimit, shouldFail)
+
+    /** Builds a `CompilationTest` with a limited level of concurrency with
+     *  maximum `i` threads
+     */
+    def limitThreads(i: Int): CompilationTest =
+      new CompilationTest(targets, times, shouldDelete, Some(i), shouldFail)
+
+    /** Builds a `CompilationTest` where the executed test is expected to fail
+     *
+     *  This behaviour is mainly needed for the tests that test the test suite.
+     */
+    def expectFailure: CompilationTest =
+      new CompilationTest(targets, times, shouldDelete, threadLimit, true)
+
+    /** Delete all output files generated by this `CompilationTest` */
+    def delete(): Unit = targets.foreach(t => delete(t.outDir))
+
+    private def delete(file: JFile): Unit = {
+      if (file.isDirectory) file.listFiles.foreach(delete)
+      try Files.delete(file.toPath)
+      catch {
+        case _: NoSuchFileException => // already deleted, everything's fine
+      }
+    }
+  }
+
+  /** Create out directory for directory `d` */
+  private def toCompilerDirFromDir(d: JFile, sourceDir: JFile, outDir: String): JFile = {
+    val targetDir = new JFile(outDir + s"${sourceDir.getName}/${d.getName}")
+    targetDir.mkdirs()
+    targetDir
+  }
+
+  /** Create out directory for `file` */
+  private def toCompilerDirFromFile(file: JFile, sourceDir: JFile, outDir: String): JFile = {
+    val uniqueSubdir = file.getName.substring(0, file.getName.lastIndexOf('.'))
+    val targetDir = new JFile(outDir + s"${sourceDir.getName}/$uniqueSubdir")
+    targetDir.mkdirs()
+    targetDir
+  }
+
+  /** Make sure that directory string is as expected */
+  private def checkRequirements(f: String, sourceDir: JFile, outDir: String): Unit = {
+    require(sourceDir.isDirectory && sourceDir.exists, "passed non-directory to `compileFilesInDir`")
+    require(outDir.last == '/', "please specify an `outDir` with a trailing slash")
```

Comment:
`java.io.File` internally handles forward slash translation if the platform is Windows.

I could provide more of an API for this, but felt it was approaching over-engineering...